core/commands/store/challenge/challenge.go

```diff
@@ -125,34 +125,20 @@ the challenge request back to the caller.`,
 		if err != nil {
 			return err
 		}
-		// Check if host store has this contract id
-		shardInfo, err := storage.GetShardInfoFromDatastore(n, storage.HostStoragePrefix, sh)
-		if err != nil {
-			return err
-		}
-		ctID := req.Arguments[0]
-		if ctID != shardInfo.ContractID {
-			return fmt.Errorf("contract id does not match, has [%s], needs [%s]", shardInfo.ContractID, ctID)
-		}
-		if !shardHash.Equals(shardInfo.ShardHash) {
-			return fmt.Errorf("datastore internal error: mismatched existing shard hash %s with %s",
-				shardInfo.ShardHash.String(), shardHash.String())
-		}
 		chunkIndex, err := strconv.Atoi(req.Arguments[3])
 		if err != nil {
 			return err
 		}
 		nonce := req.Arguments[4]
 		// Get (cached) challenge response object and solve challenge
-		sc, err := shardInfo.GetChallengeResponseOrNew(req.Context, n, api, fileHash, false, 0)
+		sc, err := storage.NewStorageChallengeResponse(req.Context, n, api, fileHash, shardHash, "", true, 0)
```

Comment:
Must use `false` as the `init` param. true means adding & pinning everything again.

Reply:
ok, on second thought, it needs to be `shardInfo.GetChallengeResponseOrNew` because
1) NewStorageChallengeResponse re-initializes all maps and info which could taka time
2) shardInfo.GetChallengeResponseOrNew uses a cached version of challenge whenever possible (no more dag initialization)

However.. we will have the problem when different contracts contain the same shard hashes.
Let's discuss.

Reply:
okay i'll read the code again.